File: 백준/Silver/18870. 좌표 압축/좌표 압축.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        long[] arr = new long[N];
        long[] copy = new long[N];
        HashMap<Long, Integer> map = new HashMap<>();
        for(int i = 0; i < N; i++) {
            arr[i] = Long.parseLong(st.nextToken());
            copy[i] = arr[i];
        }
        Arrays.sort(copy);
        int idx = 0;
        for(int i = 0; i < N; i++) {
            if(!map.containsKey(copy[i])) {
                map.put(copy[i], idx++);
            }
        }
        for(int i = 0; i < N; i++) {
            sb.append(map.get(arr[i])).append(' ');
        }
        System.out.println(sb.toString().trim());
    }
}

File: 백준/Silver/18870. 좌표 압축/좌표 압축.js
const [[N], arr] = require('fs').readFileSync('/dev/stdin').toString().split('\n').map(line => line.trim().split(' ').map(Number));
const tmp = [...new Set(arr)].sort((a, b) => a - b);
const map = new Map();
tmp.forEach((val, idx) => {map.set(val, idx)});
const ans = arr.map(val => map.get(val));
console.log(ans.join(' '));